design/host-discovery-controller/host-discovery.md

```diff
@@ -0,0 +1,325 @@
+provisioned automatically.
+
+## Design Details
+
+When Ironic is configured to support its existing discovery feature,
+unknown hosts that boot the Ironic agent will have their hardware
+details registered with `ironic-inspector` and have a node created in
+the Ironic database. Typically this happens via the PXE boot
+configuration, but in the future it could happen in other ways such as
+having a bootable ISO image with the agent on it. How the agent ends
+up on the host should not matter in the rest of the workflow.
+
+The data collected for an unknown host includes the same inspection
+data available for any other host, but does not include the BMC
+credentials. We can use the data to partially create a `BareMetalHost`
+resource for the host by polling the Ironic API to compare the list of
+nodes it knows about with the list of host resources. A new
+`host-discovery-controller` will manage this polling, using a new
+`BareMetalHostDiscovery` API to control it.
+
+The `BareMetalHostDiscovery` API will include a `resourceNameTemplate`
+field to hold instructions for building a resource name for a new
+node, including a prefix, suffix, and enum for selecting known
+hardware details.
+
+For a host with name `"the-host-name"`, this specification
+
+```yaml
+resourceNameTemplate:
+  prefix: "string-literal1-"
+  suffix: "-string-literal2"
+  hardwareDetails: hostname
+```
+
+would produce a resource name
+`"string-literal1-the-host-name-string-literal2"`.
+
+The `hardwareDetails` field is required. The `prefix` and `suffix`
+fields are ignored if not provided.
+
+The `hardwareDetails` enum can be one of:
+
+- `hostname` -- Using the discovered hostname for the host.
+- `ip` -- The IP address of the first NIC for the host with `.`
+  replaced by `-`.
+- `serial-number` -- Using the hardware serial number for the host.
+- `boot-mac` -- Using the boot MAC address of the host with `:`
+  replaced by `-`.
+- `provisioning-id` -- The identifier assigned by the underlying
+  provisioning tool when the host was registered automatically.
+
+Changing the a `BareMetalHostDiscovery` resource will have no effect
+on previously discovered host resources.
+
+For each host known to Ironic, the `host-discovery-controller` will
+look for a `BareMetalHost` resource with the same `bootMACAddress`. If
+any host resource is found matching the MAC, no further action will be
+taken.
+
+For any host known to Ironic but not represented by a `BareMetalHost`
+resource, the `host-discovery-controller` will create a
+`BareMetalHost` resource with the `bootMACAddress` filled in based on
+the MAC of the first discovered NIC and using a name built from the
+template values and with an annotation containing the hardware details
+and provisioner ID that should be added to the status fields.
+
+When a user adds BMC credentials to a host in the `Discovered` state
+today it moves to `Registering` and the `Provisioner` is responsible
```

Comment:
Ironic can now be configured to provision a node without having credentials, see my demo https://youtu.be/A9Q8qrEhyh0. I don't suggest we do it by default, but it can be a handy option.